ls1intum/Artemis
# Highlight plagiarism matches

### Checklist

### Motivation and Context
To make the inspection of similar student submissions even easier, this PR adds the functionality to highlight detected matches between submissions.

**Note**: The highlighting introduced with this PR only works for *text* and *programming* exercises.

### Description
Before we display the student submissions in the split view, we use the information provided by JPlag about the exact matches to insert a token before and right after a detected match.

### Steps for Testing

To properly test this PR, you need both a text and a programming exercise with multiple similar submissions.

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to the Exercises of a course
4. Select a text or programming exercise
5. Click 'Check Plagiarism'

For programming exercises, generate a JPlag report. Then, uncheck 'Generate JPlag Report', click 'Check Plagiarism' again, and inspect the results in the split view. Make sure the Artemis highlighting matches the JPlag highlighting:

<img width="1680" alt="Screenshot 2020-12-29 at 16 56 14" src="https://user-images.githubusercontent.com/21973908/103296766-16035900-49f7-11eb-94b4-44a898ad5f00.png">
<img width="1680" alt="Screenshot 2020-12-29 at 17 12 09" src="https://user-images.githubusercontent.com/21973908/103297634-06850f80-49f9-11eb-94be-fc2ddd02b966.png">

For text exercises, we can't compare the highlighting to the JPlag report. Just make sure you can see highlighted matches at all and that they roughly "make sense".

### Test Coverage
text-submission-viewer.component.ts 84.44%
plagiarism-sidebar.component.ts 82.14%

### Screenshots
<img width="1180" alt="Screenshot 2020-12-29 at 16 52 15" src="https://user-images.githubusercontent.com/21973908/103296683-d9376200-49f6-11eb-9fb5-130a4d22b871.png">
